### PR TITLE
Transformation matrix class - preserve transformations during reload

### PR DESCRIPTION
**Pull request by Oekn5w**
Corresponding feature request: #4401 (sidenoted in #4269, #4826 )

As it turned out, the necessary modifications span many classes, predominantly Model and its subclasses , the IO classes and the admesh utility.

The UI doesn't change apart from the option added to the reload preferences, where you can choose whether you want to try and preserve the transformations or not.

All the transformations are now normalized to result in a general purpose matrix transformation. The volumes' meshes are changed as before, but the volumes' TransformationMatrix property keep track of the transformations applied to their corresponding meshes.

On reload, the meshes are reloaded and the matrices applied when the corresponding options are selected.

As all geometric manipulations now result in matrices, those are used in the undo / redo stack as opposed to (multiple) rotations / scales / translations.

This PR is rounded off by a test file for the new class.

### CODE REVIEW

**Comment by Oekn5w** [comment]
I don't know why the 64-bit inverse test passes, but fails in 32-bit.

**Comment by Oekn5w** [comment]
@lordofhyphens thoughts? Is this PR too big? I didn't really think it would make sense to just add the class without adding the base functionality.

**Comment by lordofhyphens** [comment]
Sorry, got distracted and will look at it in a bit

File: xs/src/admesh/util.c [hunk 1]
```diff
@@ -185,24 +185,74 @@ void calculate_normals(stl_file *stl) {
...
     return;
   for (i_face = 0; i_face < stl->stats.number_of_facets; ++ i_face) {
     stl_vertex *vertices = stl->facet_start[i_face].vertex;
     for (i_vertex = 0; i_vertex < 3; ++ i_vertex) {
       stl_vertex* v_dst = &vertices[i_vertex];
-      stl_vertex  v_src = *v_dst;
-      v_dst->x = trafo3x4[0] * v_src.x + trafo3x4[1] * v_src.y + trafo3x4[2]  * v_src.z + trafo3x4[3];
-      v_dst->y = trafo3x4[4] * v_src.x + trafo3x4[5] * v_src.y + trafo3x4[6]  * v_src.z + trafo3x4[7];
-      v_dst->z = trafo3x4[8] * v_src.x + trafo3x4[9] * v_src.y + trafo3x4[10] * v_src.z + trafo3x4[11];
+      double v_src_x = (double)(v_dst->x);
```
Comment (lordofhyphens) [inline xs/src/admesh/util.c]: Was there a reason you needed to modify admesh? This is technically code external to the Slic3r project and proposed changes should also be pushed upstream.

**Comment by lordofhyphens** [comment]
Thank you very much for the work; I've done some UI testing of it and it looks fine to me. Sorry it took me this long to get to it.

File: xs/t/25_transformationmatrix.t [hunk 2]
```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/perl
```
Comment (lordofhyphens) [inline xs/t/25_transformationmatrix.t]: Can you port this to the C++ tests after merging?

File: xs/src/admesh/util.c (hunk 1 above)
Comment (Oekn5w) [reply xs/src/admesh/util.c]: as this is now the remaining function that is called in admesh, I wanted it it to be efficient wrt the checks done on each transform (size, volume, faces) and to have and option to directly assign a transformed mesh. I'll look into purposing the corresponding upstream change.

File: xs/t/25_transformationmatrix.t (hunk 2 above)
Comment (Oekn5w) [reply xs/t/25_transformationmatrix.t]: sure, I'll open another PR when I'm done

**Pull request merged**